pkg/reconciler/pipelinerun/affinity_assistant_test.go

```diff
@@ -88,6 +97,280 @@ func TestCreateAndDeleteOfAffinityAssistant(t *testing.T) {
 	}
 }
 
+// TestCreateAffinityAssistantWhenNodeIsCordoned tests an existing Affinity Assistant can identify the node failure and
+// can migrate the affinity assistant pod to a healthy node so that the existing pipelineRun runs to competition
+func TestCreateAffinityAssistantWhenNodeIsCordoned(t *testing.T) {
+	ctx := context.Background()
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	c := Reconciler{
+		KubeClientSet: fakek8s.NewSimpleClientset(),
+		Images:        pipeline.Images{},
+	}
+
+	// create affinity assistant in a cluster
+	err := c.createOrUpdateAffinityAssistants(ctx, testPipelineRun.Spec.Workspaces, testPipelineRun, testPipelineRun.Namespace)
+	if err != nil {
+		t.Errorf("unexpected error from createOrUpdateAffinityAssistants: %v", err)
+	}
+
+	// get the affinity assistant just created
+	expectedAffinityAssistantName := getAffinityAssistantName(workspaceName, testPipelineRun.Name)
+
+	// assume the necessary pod is created, update the affinity assistant with the number of pods in readyReplicas
+	var replaceReplicasPatchBytes []byte
+	replaceReplicasPatchBytes, err = json.Marshal([]jsonpatch.JsonPatchOperation{{
+		Operation: "replace",
+		Path:      "/status/readyReplicas",
+		Value:     int32(1),
+	}})
+	_, err = c.KubeClientSet.AppsV1().StatefulSets(testPipelineRun.Namespace).Patch(ctx, expectedAffinityAssistantName, types.JSONPatchType, replaceReplicasPatchBytes, metav1.PatchOptions{})
```

Comment:
The pattern we seem to use most frequently in tests is to seed the test data with test objects; would that work here? e.g. https://github.com/tektoncd/pipeline/blob/d6c0f69487ced7305725e3b741fe161c4d880971/test/controller.go#L169

We'd just have to add statefulsets and nodes to test data.

I think this would be helpful because then we can have 2 separate tests: 1 for creating the affinity assistant statefulset, and then one for testing the behavior of the function when the statefulset has 1 ready replica.

Reply:
> The pattern we seem to use most frequently in tests is to seed the test data with test objects; would that work here? 

that's a great idea, we could certainly add `statefulsets` and `nodes` to the test data. 

But I am having hard time imagining it here in these test cases.

[SeedTestData](https://pkg.go.dev/github.com/tektoncd/pipeline/test#SeedTestData) returns a list of `fakeClients`. 

So this can be changed:

```
c.KubeClientSet.AppsV1().StatefulSets(...
``` 

to:

```
c, _ := test.SeedTestData(t, ctx, d)
c.Kube.AppsV1().StatefulSets()
```

which is in the end all same but these tests are working with the `pipelinerun.Reconciler` to test a member function `createOrUpdateAffinityAssistants`. How do we test a member function without direct instance of `reconciler`. 

The `pipelinerun.go` has about 13 such member functions `func (c *Reconciler) ` without any unit tests. May be we have an opportunity to design something here which can help with the unit tests in future.